be/src/vec/columns/column_dictionary.h

```diff
@@ -0,0 +1,381 @@
+        }
+
+        if (!is_dict_code_converted()) {
+            for (size_t i = 0; i < size(); ++i) {
+                codes[i] = dict.convert_code(codes[i]);
+            }
+            dict_code_converted = true;
+        }
+    }
+
+    void sort_dict() {
+        dict.sort();
+        dict_sorted = true;
+    }
+
+    class Dictionary {
+    public:
+        Dictionary() = default;
+
+        void reserve(size_t n) {
+            dict_data.reserve(n);
+            inverted_index.reserve(n);
+        }
+
+        inline void insert_value(StringValue& value) {
+            dict_data.push_back_without_reserve(value);
+            inverted_index[value] = inverted_index.size();
+        }
+
+        inline T find_code(const StringValue& value) const {
+            auto it = inverted_index.find(value);
+            if (it != inverted_index.end()) {
+                return it->second;
+            }
+            return -1;
+        }
+
+        inline T find_bound_code(const StringValue& value, bool lower, bool eq) const {
+            if (lower) {
+                return std::lower_bound(dict_data.begin(), dict_data.end(), value) - dict_data.begin() - eq;
+            } else {
+                return std::upper_bound(dict_data.begin(), dict_data.end(), value) - dict_data.begin() + eq;
+            }
+        }
+
+        inline phmap::flat_hash_set<T> find_codes(const phmap::flat_hash_set<StringValue>& values) const {
+            phmap::flat_hash_set<T> code_set;
+            for (const auto& value : values) {
+                auto it = inverted_index.find(value);
+                if (it != inverted_index.end()) {
+                    code_set.insert(it->second);
+                }
+            }
+            return code_set;
+        }
+
+        inline StringValue& get_value(T code) { return dict_data[code]; }
+
+        void clear() {
+            dict_data.clear();
+            inverted_index.clear();
+            code_convert_map.clear();
+        }
+
+        void sort() {
+            size_t dict_size = dict_data.size();
+            std::sort(dict_data.begin(), dict_data.end(), comparator);
+            for (size_t i = 0; i < dict_size; ++i) {
+                code_convert_map[inverted_index.find(dict_data[i])->second] = (T)i;
+                inverted_index[dict_data[i]] = (T)i;
+            }
+        }
+
+        inline T convert_code(const T& code) const { return code_convert_map.find(code)->second; }
+
+        size_t byte_size() { return dict_data.size() * sizeof(dict_data[0]); }
+
+    private:
+        struct HashOfStringValue {
+            size_t operator()(const StringValue& value) const {
+                return HashStringThoroughly(value.ptr, value.len);
+            }
+        };
+
+        StringValue::Comparator comparator;
+        // dict code -> dict value
+        DictContainer dict_data;
+        // dict value -> dict code
+        phmap::flat_hash_map<StringValue, T, HashOfStringValue> inverted_index;
```

Comment:
Why use a generic type for integer value? Do you meat to support int8, int32, int64, etc?

Reply:
Because for low cardinality columns, int8 and int16 are often enough. But now the stored encoding is int32, so now the code uses int32 by default. Template is used to think that it can be optimized in the future. If you use less than 32, use 8 and 16.